### PR TITLE
Dynamic logdrains

### PR DESCRIPTION
**Pull request by sheax0r**
This pull request implements the log-iss side of a dynamic credential solution.

Credentials are retrieved from a redis server as follows:

* Periodically, all secrets are retrieved using `HGETALL`.
* We create a new set of credentials based on these secrets.
* If the new set of credentials is different from the old set, we replace the old set with the new one.

### Notes:

This PR removes the "token-scrubbing" functionality implemented in https://github.com/heroku/log-iss/pull/43. In future, such a solution will not be possible, and it is only possible right now for tokens that continue to be defined in an environment variable. In future, tokens will not be accessible to log-iss at all - only HMACs based on tokens will be available via Redis.

I have also copied the [basic_auth](https://github.com/heroku/authenticater/blob/master/basic_auth.go) code from authenticater, and I think we can just remove authenticater as a dependency. This code is not complicated, and we can implement a better-performing, cleaner solution if we have access to some fields that `authenticater` itself hides -- specifically [BasicAuth.env](https://github.com/heroku/authenticater/blob/16ff6099109815e5523480d377232dc6021ba3cf/basic_auth.go#L14). Copying this code also allows us to add more metrics, which can help with cred-roll tracking.

### CODE REVIEW

**Comment by sheax0r** [comment]
I think that the rate limits described [here](https://docs.aws.amazon.com/secretsmanager/latest/userguide/reference_limits.html) are going to make it necessary to change this implementation.

With lots of passwords and/or lots of log-iss processes, it will be far too easy to hit our limits. I'll go in and alter this to read from a redis key instead.

File: cmd/forwarder/config.go [hunk 1]
```diff
@@ -31,8 +29,19 @@ type IssConfig struct {
...
 	MetricsRegistry           metrics.Registry
-	tokenMap                  map[string]string
-	tokenMapOnce              sync.Once
+}
+
+type AuthConfig struct {
+	RedisUrl        string        `env:"REDIS_URL,required"`
+	RedisKey        string        `env:"REDIS_KEY",required`
+	RefreshInterval time.Duration `env:"TOKEN_REFRESH_INTERVAL,default=1m,strict"`
+	Tokens          string        `env:"TOKEN_MAP,required"`
```
Comment (jmervine) [inline cmd/forwarder/config.go]: @sheax0r Is the plan to store the old `TOKEN_MAP` string format in Redis? If so, what's the motivation behind that, over storing user/token pairs?

Comment (sheax0r) [reply cmd/forwarder/config.go]: @jmervine the plan is not to store the old `TOKEN_MAP` format in redis. The plan is to store a redis hash using [HGETALL](https://redis.io/commands/hgetall). Each key in the hash is a username, and each value is a json array of passwords that are valid for the user. This allows us to store a data structure that is equivalent to the old `TOKEN_MAP` value, but requires less custom string-manipulation (json has libs to manipulate it; the `TOKEN_MAP` string requires custom code and custom escape sequences).

TL;DR: This format gives us the same power as `TOKEN_MAP`, but is easier to read and write, and leverages native redis capabilities.

File: cmd/forwarder/auth.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-redis/redis"
+	metrics "github.com/rcrowley/go-metrics"
+	log "github.com/sirupsen/logrus"
+)
+
+func newAuth(config AuthConfig, registry metrics.Registry) (*BasicAuth, error) {
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Generally commentary on this function ... You may want to have two functions, one for handling auth based on config.Tokens and another for auth based on redis. And then wrap them in a single function.

File: cmd/forwarder/auth.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+		)
+
+		// Refresh once at the start.
+		// Ignore errors here - if redis is down, we don't want the process to fail to start.
+		_, _ = refreshAuth(result, client, config.HmacKey, config.RedisKey, config.Tokens)
+
+		// Refresh forever
+		ticker := time.NewTicker(config.RefreshInterval)
+		go func() {
+			changed := false
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Suggestion: Prefer `var changed bool` as false is the zero value of a bool.

File: cmd/forwarder/auth.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+			for _ = range ticker.C {
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Suggestion, allowing you to remove the above `_ = refreshAuth... ` stuff

```go
ticker := time.NewTicker(config.RefreshInterval)
for ; true; <-ticker.C {
    _, err := refreshAuth(result, client, config.HmacKey, config.RedisKey, config.Tokens)
  ...
}
```

File: cmd/forwarder/auth.go [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+				_, err := refreshAuth(result, client, config.HmacKey, config.RedisKey, config.Tokens)
+				if err == nil {
+					pSuccesses.Inc(1)
+					if changed {
+						pChanges.Inc(1)
+					}
+				} else {
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: put a `continue` above this line and you can do away with the `else`.

File: cmd/forwarder/auth.go [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+			},
+		)
+
+		// Refresh once at the start.
+		// Ignore errors here - if redis is down, we don't want the process to fail to start.
+		_, _ = refreshAuth(result, client, config.HmacKey, config.RedisKey, config.Tokens)
+
+		// Refresh forever
+		ticker := time.NewTicker(config.RefreshInterval)
+		go func() {
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Suggestion: Consider making this whole thing a named function and the `go func(args)` that.

File: cmd/forwarder/auth.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+			changed := false
+			for _ = range ticker.C {
+				_, err := refreshAuth(result, client, config.HmacKey, config.RedisKey, config.Tokens)
+				if err == nil {
+					pSuccesses.Inc(1)
+					if changed {
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Also, I don't see `changed` being set to `true` anywhere.

File: cmd/forwarder/auth.go [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+	val := client.HGetAll(redisKey)
+	r, err := val.Result()
+	if err != nil {
+		return false, err
+	}
+
+	for k, v := range r {
+		var arr []Credential
+		err = json.Unmarshal([]byte(v), &arr)
+
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Please drop newline

File: cmd/forwarder/auth.go [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+		// Refresh once at the start.
+		// Ignore errors here - if redis is down, we don't want the process to fail to start.
+		_, _ = refreshAuth(result, client, config.HmacKey, config.RedisKey, config.Tokens)
+
+		// Refresh forever
+		ticker := time.NewTicker(config.RefreshInterval)
+		go func() {
+			changed := false
+			for _ = range ticker.C {
+				_, err := refreshAuth(result, client, config.HmacKey, config.RedisKey, config.Tokens)
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: I think you meant for this to be `changed, err := ....`. If so you don't need to declare `changed` above.

File: cmd/forwarder/auth.go [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+				if err == nil {
+					pSuccesses.Inc(1)
+					if changed {
+						pChanges.Inc(1)
+					}
+				} else {
+					fmt.Printf("Unable to refresh credentials: %s", err)
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: The rest of log-iss uses logrus, so that should be used here too.

File: cmd/forwarder/auth.go [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+		return result, err
+	}
+
+	pChanges := metrics.GetOrRegisterCounter("log-iss.auth_refresh.changes", registry)
+	pFailures := metrics.GetOrRegisterCounter("log-iss.auth_refresh.failures", registry)
+	pSuccesses := metrics.GetOrRegisterCounter("log-iss.auth_refresh.successes", registry)
+
+	if config.RedisUrl != "" {
+		// Parse redis db out of url
+		u, err := url.Parse(config.RedisUrl)
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Instead of all of this see: https://godoc.org/github.com/go-redis/redis#ParseURL

File: cmd/forwarder/auth.go [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+// Return true if credentials changed, false otherwise.
+func refreshAuth(ba *BasicAuth, client redis.Cmdable, hmacKey string, redisKey string, config string) (bool, error) {
+	// Start out using the strings from config
+	nba, err := NewBasicAuthFromString(config, hmacKey, ba.registry)
+	if err != nil {
+		return false, err
+	}
+
+	// Retrieve all secrets and construct a string in the format expected by BasicAuth
+	val := client.HGetAll(redisKey)
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: You probably need a nil check here because `client.HGetAll()` returns a `*StringStringMapCmd` and I can't tell from the docs if it ensures that it's non nil.

File: cmd/forwarder/auth.go [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+	}
+
+	for k, v := range r {
+		var arr []Credential
+		err = json.Unmarshal([]byte(v), &arr)
+
+		if err != nil {
+			return false, err
+		}
+		nba.creds[k] = arr
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Do creds from redis take precedence over creds from the environment? If so, 👍 , if not then the ordering needs to be reversed.

File: cmd/forwarder/auth.go [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+	ba := NewBasicAuth(registry, hmacKey)
+
+	// If the string is empty, that's allowed.
+	if creds == "" {
+		return ba, nil
+	}
+
+	for _, u := range strings.Split(creds, "|") {
+		uparts := strings.SplitN(u, ":", 2)
+		if len(uparts) != 2 || len(uparts[0]) == 0 || len(uparts[1]) == 0 {
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Suggestion prefer: `uparts[0] == "" || uparts[1] == ""`

File: cmd/forwarder/auth.go [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+			return ba, fmt.Errorf("Unable to create credentials from '%s'", u)
+		}
+
+		value := uparts[1]
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Suggestion: Use `uparts[1]` on the next line and do away with value.

File: cmd/forwarder/auth.go [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+		ba.AddPrincipal(uparts[0], hmacEncode(hmacKey, value), "env")
+	}
+	return ba, nil
+}
+
+func hmacEncode(key string, value string) string {
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: suggestion rename `value -> message` and below use those in 171/172 doing away with 169 & 170.

File: cmd/forwarder/auth.go [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+	return &BasicAuth{
+		creds:    make(map[string][]Credential),
+		hmacKey:  hmacKey,
+		registry: registry,
+	}
+}
+
+// AddPrincipal add's a user/password combo to the list of valid combinations
+func (ba *BasicAuth) AddPrincipal(user string, hmac string, stage string) {
+	ba.Lock()
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Suggestion: `defer ba.Unlock()` below this line and remove it from line 192

File: cmd/forwarder/auth.go [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+	u, existed := ba.creds[user]
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Suggestion: s/existed/exists/ or s/existed/ok/

File: cmd/forwarder/auth.go [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+				counter := metrics.GetOrRegisterCounter(countName, ba.registry)
+				counter.Inc(1)
+				return true
+			}
+		}
+		log.WithFields(log.Fields{"ns": "auth", "at": "failure", "user": user}).Info()
+		countName := fmt.Sprintf("log-iss.auth.%s.failures", user)
+		counter := metrics.GetOrRegisterCounter(countName, ba.registry)
+		counter.Inc(1)
+	} else {
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Suggestion: return above the else, remove the else, dedent everything.

Comment (freeformz) [reply cmd/forwarder/auth.go]: other suggestion:
```go
credentials, exists := ba.creds[user]
if !exists {
  log.WithFields(log.Fields{"ns": "auth", "at": "failure", "user": user}).Info()
  return false
}

for _, c := range credentials {
  ....
}
```

File: cmd/forwarder/auth_test.go [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+		"env creds can be overridden": {
+			auth:          defaultCreds(),
+			client:        overrideRedis(),
+			config:        "user:password",
+			expectedCreds: overrideCreds(),
+			expectChanged: true,
+		},
+	}
+
+	for name, test := range tests {
```
Comment (freeformz) [inline cmd/forwarder/auth_test.go]: Can you not use t.Run() here?

File: cmd/forwarder/auth_test.go [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+				RedisUrl:        "redis://localhost:6379/0",
+				RedisKey:        "key",
+			},
+			success: true,
+		},
+	}
+
+	registry := metrics.NewRegistry()
+
+	for name, test := range tests {
```
Comment (freeformz) [inline cmd/forwarder/auth_test.go]: Same t.Run() question.

File: cmd/forwarder/auth_test.go [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+			authenticated: false,
+		},
+	}
+
+	auth, err := NewBasicAuthFromString("user:password", "hmacKey", metrics.NewRegistry())
+	if err != nil {
+		panic(err.Error())
+	}
+
+	for name, test := range tests {
```
Comment (freeformz) [inline cmd/forwarder/auth_test.go]: same t.Run() question.

File: cmd/forwarder/credential.go [hunk 23]
```diff
@@ -0,0 +1,6 @@
+package main
+
+type Credential struct {
```
Comment (freeformz) [inline cmd/forwarder/credential.go]: Godoc please

File: cmd/hash/main.go [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+	message := []byte(value)
+	hash := hmac.New(sha512.New, secret)
+	hash.Write(message)
+	return hex.EncodeToString(hash.Sum(nil))
+}
+
+func main() {
+	var config AuthConfig
+	err := envdecode.Decode(&config)
+
```
Comment (freeformz) [inline cmd/hash/main.go]: Please drop the newline.

File: cmd/forwarder/auth.go [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+	"github.com/go-redis/redis"
+	metrics "github.com/rcrowley/go-metrics"
+	log "github.com/sirupsen/logrus"
+)
+
+func newAuth(config AuthConfig, registry metrics.Registry) (*BasicAuth, error) {
+	var err error
+
+	if config.RedisUrl != "" && config.RedisKey == "" {
+		err = errors.New("RedisKey must be set if RedisUrl is set")
```
[the remark this answers is not in the document]
Comment (sheax0r) [reply cmd/forwarder/auth.go]: To answer your question: Because writing extraneous assignment lines is something that devs do when they are new to a language. 

I'll take your suggestion and return the error without the assignment.

File: cmd/forwarder/auth_test.go (hunk 20 above)
Comment (sheax0r) [reply cmd/forwarder/auth_test.go]: @freeformz I found this guide to writing table-based tests in go: https://medium.com/agrea-technogies/basic-testing-patterns-in-go-d8501e360197

It didn't mention `t.Run()`, unlike the docs here: https://golang.org/pkg/testing/

So - yes, I can use `t.Run()` and I'll change the code to do so.  I followed a guide that made no mention of this.

File: cmd/forwarder/auth_test.go (hunk 21 above)
Comment (sheax0r) [reply cmd/forwarder/auth_test.go]: Same answer.

File: cmd/forwarder/auth_test.go (hunk 22 above)
Comment (sheax0r) [reply cmd/forwarder/auth_test.go]: Same answer.

File: cmd/forwarder/credential.go (hunk 23 above)
Comment (sheax0r) [reply cmd/forwarder/credential.go]: @freeformz I have added godoc and made Credential private, moving it into the `auth.go` package. Should I *remove* the godoc now that it's private?

File: cmd/forwarder/auth.go (hunk 2 above)
Comment (sheax0r) [reply cmd/forwarder/auth.go]: @freeformz I'm working on this one.

Comment (sheax0r) [reply cmd/forwarder/auth.go]: @freeformz so, I have some thoughts here. 

I originally started out attempting to have multiple objects, each with their own `Authenticate` method. I ran into a few problems doing it. Basically:

If I have multiple invocations of an `Authenticate(r *http.Request)` function (eg: one for creds that come from the environment and another for creds from redis), then under-the-hood, each will need to invoke `hmacEncode` on the input password. This is expensive.

There are ways to mitigate this expense: 

1. I could cache the result of hmacEncode in thread local storage. But if I do this, I must be careful: A brute-force password attack could cause us to cache the results of invalid passwords forever. I need to ensure the cache has an upper size bound, and evicts negative results eventually. I'm not sure what to use to for that. I'm also not sure how to do thread-local storage in go at this point.

2. I could implement a different interface like `Authenticate(user string, passwordHash string), and just have an object that contains a map of strings to a list of `credential` structs.  However, with that implementation I'm starting to feel like there's very little logic in the methods specific to this struct, and it's really just a wrapper around the map - in which case I feel like it's over-engineered and not adding alot of value to the code.

Do you have any suggestions for how I should proceed? If I don't do this now, I think it's unlikely that I will ever do it.

Comment (sheax0r) [reply cmd/forwarder/auth.go]: @freeformz my preference for right now, sadly, is to not do it, unless you feel strongly. If I don't deploy this during my day tomorrow, I'm unlikely to be able to deploy it until next week.

Comment (freeformz) [reply cmd/forwarder/auth.go]: Sorry, I may not have been concrete enough in my comment. I was trying to say:
```go
func newAuth(...) (*BasicAuth, error) {
// config checks
....

result, err := NewBasicAuthFromString(config.Tokens, config.HmacKey, registry)
if err != nil {
  return result, err
}

if config.RedisUrl != "" {
  go refreshAuthFromRedis(result, registry, config.RedisUrl, config.HmacKey, config.RedisKey, config.Tokens)
}
return result, err
```

Basically fully split it up into two functions, each focused on their own source. The one that deals with config.Tokens source already exists in the form of NewBasicAuthFromString.

File: cmd/forwarder/auth_test.go (hunk 20 above)
Comment (freeformz) [reply cmd/forwarder/auth_test.go]: I think that guide predated t.Run()

File: cmd/forwarder/credential.go (hunk 23 above)
Comment (freeformz) [reply cmd/forwarder/credential.go]: I'd keep it. docs are good.

File: cmd/forwarder/auth.go (hunk 13 above)
Comment (sheax0r) [reply cmd/forwarder/auth.go]: @freeformz hey - i missed some convos from before, i think github was hiding them from me. i'm looking for them today. 

regarding this comment: yes, creds from redis take precedence, by design.

File: cmd/forwarder/auth.go (hunk 7 above)
Comment (sheax0r) [reply cmd/forwarder/auth.go]: Indeed, it wasn't. This will be fixed next commit.

File: cmd/forwarder/auth.go (hunk 14 above)
Comment (sheax0r) [reply cmd/forwarder/auth.go]: I copied this wholesale from [here](https://github.com/heroku/authenticater/blob/master/basic_auth.go#L30-L32).

I tested this code out [here](https://play.golang.org/p/VSNxKKXAcyR) and I believe that using this suggestion will segfault if it encounters malformed input.

**Comment by sheax0r** [comment]
@freeformz I have responded to every comment I can find (although I'm still not 100% sure I'm seeing all conversations). Please let me know what you think of the latest changes, and if you have further recommendations. 

Thanks for this, it is very helpful!

File: cmd/forwarder/auth.go (hunk 17 above)
Comment (sheax0r) [reply cmd/forwarder/auth.go]: @freeformz Agree, your suggestion is better. This code was also copied from [here](https://github.com/heroku/authenticater/blob/master/basic_auth.go#L41-L49)

File: cmd/forwarder/auth.go [hunk 26]
```diff
@@ -0,0 +1,212 @@
...
+		changed, err := auth.refresh(client, config.HmacKey, config.RedisKey, config.Tokens)
+		if err == nil {
+			pSuccesses.Inc(1)
+			if changed {
+				pChanges.Inc(1)
+			}
+			continue
+		}
+		log.WithFields(log.Fields{"ns": "auth", "at": "error", "refresh": true, "message": err.Error()}).Info()
+		fmt.Printf("Unable to refresh credentials: %s", err)
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: Please drop the `fmt.Printf`.

File: cmd/forwarder/auth.go [hunk 27]
```diff
@@ -0,0 +1,212 @@
...
+	// Parse redis db out of url
+	opt, err := redis.ParseURL(config.RedisUrl)
+	if err != nil {
+		return result, err
+	}
+	client := redis.NewClient(opt)
+
+	// Refresh once at the start.
+	// Ignore errors here - if redis is down, we don't want the process to fail to start.
+	_, _ = result.refresh(client, config.HmacKey, config.RedisKey, config.Tokens)
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: This can be dropped. The way `startRefresh` is written means auth.refresh will happen almost immediately.

File: cmd/forwarder/auth.go [hunk 28]
```diff
@@ -0,0 +1,212 @@
...
+	"github.com/go-redis/redis"
+	metrics "github.com/rcrowley/go-metrics"
+	log "github.com/sirupsen/logrus"
+)
+
+// credentials are used by basic auth and include the hash of a valid password, plus
+// a "stage" string which is used to emit metrics that are useful when managing credrolls, so that
+// we can track whether or not deprecated passwords are still in use.
+type credential struct {
+	Stage string `json:stage`
```
Comment (freeformz) [inline cmd/forwarder/auth.go]: There aren't valid struct tags: https://play.golang.org/p/nELuf2qHQU7

Comment (sheax0r) [reply cmd/forwarder/auth.go]: @freeformz interesting. i wonder how it's working? ... i'll investigate tomorrow. 

seriously though, i think it *is* working.

**Comment by sheax0r** [comment]
Btw @freeformz I'm going to deploy this tomorrow, unless you have blocking objections.

File: cmd/forwarder/auth.go (hunk 28 above)
Comment (freeformz) [reply cmd/forwarder/auth.go]: It's working, just using `Stage` and `Hmac` as the keys vs. `stage` and `hmac`

**Comment by freeformz** [comment]
SGTM

File: cmd/forwarder/auth.go (hunk 28 above)
Comment (sheax0r) [reply cmd/forwarder/auth.go]: TIL `go vet`.

Comment (sheax0r) [reply cmd/forwarder/auth.go]: Also, `json.Unmarshal` will match public members with appropriately named json hash keys even if there aren't any struct tags.

**Comment by sheax0r** [comment]
@freeformz well, I know that if I don't fix things at *review time*, I'm unlikely to ever fix them. So I was waiting, but with an eventual hard stop date of "today", which thanks to your tireless review efforts, we hit.

Thanks again for all of the tips. I will try to internalize them, and hopefully require fewer fixes the next time I call on you!

**Comment by freeformz** [comment]
👍 Happy to help, sorry for the long turnarounds in some cases. I'm just getting back to coding more, so I'm trying to be more responsive to github notifications.
